docs: updated contributing guidelines (#1067)
SwissDataScienceCenter/renku-python: CONTRIBUTING.rst

Types of Contributions
----------------------

Report issues / contacting developers

> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: make sure the linter won't complain about this - looks too long to my eyes :)

> Report bugs at our issue tracker_.

Reviewer: ```suggestion
Report bugs on our issue tracker_.
```

> If you want to submit a bug, improvement or feature suggestions feel free to open a corresponding issue on GitHub.

Reviewer: please make sure to keep consistent line length

> Look through the GitHub issues for features. Anything tagged with "feature"

Reviewer: I don't think we actually tag anything with "feature"

is open to whoever wants to implement it.

Improvement requests
~~~~~~~~~~~~~~~~~~~~

> If you see some of the things could be done in a better way, feel free to submit a issue with a improvement suggestion.

Reviewer: wording suggestion:

If you see room for improvement, please open an issue with a suggestion.

> In suggestion you might include motivation for improvement and why new solution would be improvement.

Reviewer: Wording suggestion:

Please motivate your suggestion by illustrating a problem it solves.

      $ git clone git@github.com:your_name_here/renku.git

3. Ensure you have your development environment set up. For this we encourage usage of `pipenv`

> (other virtual environment tools could be used as well) `pyenv`:

Reviewer: `pyenv` is kind of coming out of the blue for me here - aren't you talking about pipenv?

   .. code-block:: console

      $ pyenv install 3.7.5rc1
      $ cd renku/
      $ pipenv install --python  ~/.pyenv/versions/3.7.5rc1/bin/python
      $ pipenv shell

4. Create a branch for local development:

   .. code-block:: console

      $ git checkout -b <issue_number>_<short_description>

   Now you can make your changes locally.

5. When you're done making changes, check that your changes pass tests:

   .. code-block:: console

>       $ pipenv run tests

Reviewer: ```suggestion
      $ pipenv run ./run-tests.sh
```

Reviewer: Oh I didn't realize this wasn't changed - it's currently incorrect

7. Submit a pull request through the GitHub website.

Commit message guidelines
-------------------------

This project is using conventional commits style for generation of changelog upon each release. Therefore,
it's important that our commit messages convey what they do correctly. Commit message should always follow this pattern:

   $ %{type}(%{scope}): %{description}

**Type's used for describing commit's which will end up in changelog are** :code:`fix:` & :code:`feat:`.

> Please note, that fix type here is only for user facing bug fixes and not fixes on tests or CI.

Reviewer: ```suggestion
Please note that the ``fix`` type here is only for user facing bug fixes and not fixes on tests or CI.
```

For those, please use: :code:`ci:` or :code:`test:`

Full list of types which are in use:

>   * :code:`feat:` - Used for new user facing features. This should be related to one of the predefined scopes. If a scope does not exist, new scope should be proposed for a adoption.

Reviewer: ```suggestion
  * :code:`feat:` - Used for new user facing features. This should be related to one of the predefined scopes. If a scope does not exist, a new scope may be proposed.
```

>   * :code:`fix:` - Used for fixing user facing bugs. This should be related to one of the predefined scopes.

Reviewer: ```suggestion
  * :code:`fix:` - Used for fixing user-facing bugs. This should be related to one of the predefined scopes.
```

>   * :code:`chore:` - Used for change which are not user facing. Scope should be a module name in which chore occurred.

Reviewer: ```suggestion
  * :code:`chore:` - Used for changes which are not user facing. The scope should be a module name in which chore occurred.
```

>   * :code:`test:` - Used for fixing existing or adding new tests. Scope should relate to a module name which is being tested.

Reviewer: should we use ``tests``?

```suggestion
  * :code:`test:` - Used for fixing existing or adding new tests. The scope should relate to a module name which is being tested.
```

  * :code:`docs:` - Used for adding more documentation. If documentation is not related to predefined user scopes, it can be omitted.
  * :code:`refactor` - Used for changing the code structure. Scope being used here should be module name. If refactoring is across multiple modules, scope could be omitted or PR broken down into smaller chunks.

> Full list of user facing scope's which are in use:

Reviewer: ```suggestion
Full list of user facing scopes which are in use:
```

7. Submit a pull request through the GitHub website.

Commit message guidelines
-------------------------

This project is using conventional_ commits style for generation of changelog upon
each release. Therefore, it's important that our commit messages convey what they
do correctly. Commit message should always follow this pattern:

.. _conventional: https://www.conventionalcommits.org/en/v1.0.0/

   $ %{type}(%{scope}): %{description}

**Type's used for describing commit's which will end up in changelog are** :code:`fix:` & :code:`feat:`.

Please note that the :code:`fix` type here is only for user-facing bug fixes and not fixes on tests or CI.
For those, please use: :code:`ci:` or :code:`test:`

Full list of types which are in use:
  * :code:`feat:` - Used for new user-facing features. This should be related to one of the predefined scopes. If a scope does not exist, a new scope may be proposed.
  * :code:`fix:` - Used for fixing user-facing bugs. This should be related to one of the predefined scopes.
  * :code:`chore:` - Used for changes which are not user-facing. The scope should be a module name in which chore occurred.

>   * :code:`tests:` - Used for fixing existing or adding new tests. The scope should relate to a module name which is being tested.

Reviewer 2: Since we are testing a feature here the scope should come from a predefined scope too.

Author: Well, not always. Tests could be for a given utility functions which could be used across multiple scopes, that's why imho it makes more sense per module level. But I guess omitting the scope would make most sense in this case?

Reviewer 2: Yes, that also makes sense to me.

  * :code:`docs:` - Used for adding more documentation. If documentation is not related to predefined user scopes, it can be omitted.
  * :code:`refactor` - Used for changing the code structure. Scope being used here should be module name. If refactoring is across multiple modules, scope could be omitted or PR broken down into smaller chunks.

Full list of user-facing scopes which are in use:
  * :code:`graph` - Scope for describing knowledge graph which is being build with users usage of the system.
  * :code:`workflow` - Scope for describing reproducibility flow.
  * :code:`dataset` - Scope for describing datasets.
  * :code:`core` - General scope for describing all core elements of Renku project.

>   * :code:`svc` - General scope for describing interaction or operation of Renku as a service.

Reviewer 2: I prefer `service` instead of `svc`.

Reviewer: At the risk of things being more verbose, I second that

Commit message guidelines
-------------------------

> This project is using conventional commits style for generation of changelog upon

Reviewer: Maybe add a link to conventional commits?

Author: Added in #4fd255
